### PR TITLE
Tiered compilation (singlepass -> cranelift)

### PR DESCRIPTION
**Pull request by john-sharratt**
This pull request introduces the tiered compiler which will compile via `cranelift` in the background while running `singlepass` - this gives the performance benefits of a decent compiler with the great startup times of singlepass.

To use it do this...
```
john@AlienWorld:/prog/wasmer/lib/cli$ RUST_LOG=trace cargo run --release --features tiered,compiler -- run --net --disable-cache --mapdir /public:/prog/deploy/wasmer-web/dist /prog/static-web-server.wasm -- --log-level info --port 9080
    Finished release [optimized] target(s) in 0.17s
     Running `/prog/wasmer/target/release/wasmer run --net --disable-cache --mapdir '/public:/prog/deploy/wasmer-web/dist' /prog/static-web-server.wasm -- --log-level info --port 9080`
2023-03-31T05:43:46.253748Z  INFO static_web_server::logger: logging level: info
2023-03-31T05:43:46.259131Z  INFO static_web_server::server: server bound to tcp socket [::]:9080
2023-03-31T05:43:46.259843Z  INFO static_web_server::server: runtime worker threads: 4
2023-03-31T05:43:46.259908Z  INFO static_web_server::server: runtime max blocking threads: 20
2023-03-31T05:43:46.259970Z  INFO static_web_server::server: security headers: enabled=false
2023-03-31T05:43:46.260020Z  INFO static_web_server::server: auto compression: enabled=true
2023-03-31T05:43:46.260068Z  INFO static_web_server::server: directory listing: enabled=false
2023-03-31T05:43:46.260203Z  INFO static_web_server::server: directory listing order code: 6
2023-03-31T05:43:46.260304Z  INFO static_web_server::server: cache control headers: enabled=true
2023-03-31T05:43:46.260444Z  INFO static_web_server::server: basic authentication: enabled=false
2023-03-31T05:43:46.260482Z  INFO static_web_server::server: log remote address: enabled=false
2023-03-31T05:43:46.260597Z  INFO static_web_server::server: grace period before graceful shutdown: 0s
2023-03-31T05:43:46.260746Z  INFO Server::start_server{addr_str="[::]:9080" threads=4}: static_web_server::server: close time.busy=0.00ns time.idle=4.61µs
2023-03-31T05:43:46.260800Z  INFO static_web_server::server: listening on http://[::]:9080
2023-03-31T05:43:46.260857Z  INFO static_web_server::server: press ctrl+c to shut down the server
```

Looking for early feedback

### CODE REVIEW

File: lib/compiler-singlepass/src/emitter_x64.rs [hunk 1]
```diff
@@ -1471,12 +1471,24 @@ impl EmitterX64 for AssemblerX64 {
...
+            }
+            (Size::S8, Location::Imm64(imm), Size::S32, Location::GPR(dst)) => {
                 dynasm!(self ; mov Rq(dst as u8), imm as i32);
             }
             (Size::S16, Location::Imm64(imm), Size::S64, Location::GPR(dst)) => {
                 dynasm!(self ; mov Rq(dst as u8), imm as i32);
             }
+            (Size::S32, Location::Imm64(imm), Size::S64, Location::GPR(dst)) => {
+                dynasm!(self ; mov Rq(dst as u8), imm as i32);
+            }
```
Comment (syrusakbary) [inline lib/compiler-singlepass/src/emitter_x64.rs:1491]: Can we split this in other PR?

File: lib/api/src/instance.rs [hunk 2]
```diff
@@ -96,6 +103,22 @@ impl Instance {
...
+        module: &Module,
+    ) -> Result<(), InstantiationError> {
+        self._inner.reinitialize(store, module)
+    }
+
+    #[doc(hidden)]
+    #[deprecated = "this is an internal API and is subject to change"]
+    pub fn set_module(&mut self, module: &Module) {
+        self.module = module.clone();
+    }
```
Comment (syrusakbary) [inline lib/api/src/instance.rs:120]: If we want to do this should be over a trait in sys

File: lib/api/src/module.rs [hunk 3]
```diff
@@ -400,6 +400,23 @@ impl Module {
...
+        self.0.is_upgradable()
+    }
+
+    /// Tries to upgrade this module to its next tier
+    ///
+    /// Normally an upgrade operation is used to transition a module from one
+    /// tier to another. For example from `singlepass` to `cranelift`
+    pub fn try_upgrade(&self, engine: &impl AsEngineRef) -> Option<Module> {
+        self.0.try_upgrade(engine).map(|m| Self(m))
+    }
```
Comment (syrusakbary) [inline lib/api/src/module.rs:419]: If we want to do this should be over a trait in sys

Comment (john-sharratt) [reply lib/api/src/module.rs:419]: This has since been deleted

File: lib/api/src/instance.rs (hunk 2 above)
Comment (john-sharratt) [reply lib/api/src/instance.rs]: This has been deleted

File: lib/compiler-singlepass/src/emitter_x64.rs (hunk 1 above)
Comment (john-sharratt) [reply lib/compiler-singlepass/src/emitter_x64.rs:1491]: Yeah sure - that would make sense

Comment (john-sharratt) [reply lib/compiler-singlepass/src/emitter_x64.rs:1491]: https://github.com/wasmerio/wasmer/pull/3731

File: lib/api/src/module.rs (hunk 3 above)
Comment (john-sharratt) [reply lib/api/src/module.rs]: Correction its not delete - its still there - how do you recommend we create a trait for this?

**Comment by syrusakbary** [comment]
Closing this for now

**Pull request closed without merging**